Add the change event for the Cookie Store API

RupinMittal opened the pull request:
#### 5e29d991fbf6dd8193d81b5fd706b2224ed7caf2
<pre>
Add the change event for the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=260028">https://bugs.webkit.org/show_bug.cgi?id=260028</a>

Reviewed by Chris Dumez.

CookieStore now subclasses CookieChangeListener so that when an
event listener changes, it can register (or unregister) itself
as a listener. If it is registered and a cookie changed, the
relevant function (cookiesAdded or cookiesDeleted) will be called
to fire a change event.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_basic.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_delete.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_overwrite.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/httponly_cookies.https.window-expected.txt:
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt: Added.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt: Added.
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.idl:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::cookiesAdded):
(WebCore::CookieStore::cookiesDeleted):
(WebCore::CookieStore::stop):
(WebCore::CookieStore::virtualHasPendingActivity const):
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/loader/CookieJar.h:

Canonical link: <a href="https://commits.webkit.org/266850@main">https://commits.webkit.org/266850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aafa4567f05c9138315af26ca4da2d355bb35f8d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14904 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15562 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16652 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14021 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15041 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15309 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16657 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15563 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17385 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12842 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13443 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20417 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16854 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14163 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11971 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13448 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3604 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17780 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt:
```diff
@@ -1,9 +1,7 @@
+FAIL CookieStore agreed with HTTP headers agree on encoding non-ASCII cookies assert_equals: Cookie we wrote using HTTP in cookie jar expected "HTTP-🍪=🔵" but got "HTTP-ðª=ðµ"
+FAIL CookieStore set/overwrite/delete observed in HTTP headers assert_equals: Observed cookie that was deleted: number of changed cookies expected 0 but got 1
+FAIL HTTP headers agreed with CookieStore on encoding non-ASCII cookies assert_equals: HTTP cookie jar contains only cookie we set expected "🍪=🔵" but got ""
+FAIL Binary HTTP set/overwrite/delete observed in CookieStore assert_equals: Binary cookie we wrote using HTTP in cookie jar expected "HTTP-cookie=value\ufffd" but got "HTTP-cookie=valueï¿½"
```
I'm still looking into these new failures and the new failures in the other tests. Not yet 100% sure why they're happening.

RupinMittal replied:
Could this be a case of a test simply getting further than it was before?

cdumez replied:
It is probably fine to investigate this in a follow-up.

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_delete.https.window-expected.txt:
```diff
@@ -1,5 +1,3 @@
-Harness Error (TIMEOUT), message = null
-
-TIMEOUT cookieStore fires change event for cookie deleted by cookieStore.delete() Test timed out
+FAIL cookieStore fires change event for cookie deleted by cookieStore.delete() assert_equals: expected 1 but got 0
```
When a cookie is deleted, it is deleted using the CookieStore::set() function, which corresponds to a cookie being _changed_, not _deleted_. So when the Cookie Store is a listener, the change function called is `cookiesAdded` and not `cookiesDeleted`. How should we fix this? One option is that in `cookiesAdded`, if the expiry of the cookie is more than 24 hours in the past (that's what we set the expiry to be when deleting a cookie by setting it), then we can call `cookiesDeleted` for that and other such cookies. Is there a different idea that could be better? Basically, `cookiesDeleted` isn't being called at all for CookieStore and that's why this test is failing.

RupinMittal replied:
Fixed this by checking if the expiry date of the cookie is in the past. If yes, it's added to the list of deleted cookies.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+        return;
+
+    auto& document = *downcast<Document>(context);
+    if (host != document.url().host().toString())
```
This shouldn't be happening, should probably be an assertion instead.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+
+    Vector<CookieListItem> cookieList;
+    for (auto cookie : cookies)
+        cookieList.append(CookieListItem { cookie });
```
I believe you'll like need to go though the cookies to see if they have an expiration in the past. If they do, add them to the deleted list instead of the changed list.

RupinMittal replied:
Yeah, this makes sense. But what do we do with `cookiesDeleted` since it isn't being called at all?

cdumez replied:
It gets called when specific CFNetwork deletion API gets used so you still need to handle it.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -62,6 +63,10 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget, pu
+    void cookiesDeleted(const String& host, const Vector<Cookie>&) final;
+
     // ActiveDOMObject
     const char* activeDOMObjectName() const final;
```
You'll need to override ActiveDOMObject::virtualHasPendingActivity() and return true when `m_hasChangeEventListener` is set, to keep the JS wrapper alive when there is an active cookie change listener.

RupinMittal replied:
Oh I see. I'm just curious. The JS wrapper for what? And also, where does `ActiveDOMObject::virtualHasPendingActivity()` get called?

cdumez replied:
This object is exposed to Javascript so whenever the JS interacts with this object, they interacts with its JS wrapper and the wrapper basically converts JS calls into C++ calls.

It is a requirement for the JS wrapper to exist to be able to dispatch events on it. hasPendingActivity() will prevent garbageCollection of the JS wrapper as long as we return true.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+        return;
+
+    auto& document = *downcast<Document>(context);
+    if (host != document.url().host().toString())
```
Shouldn't be happening, we likely want an assertion instead.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    if (host != document.url().host().toString())
+        return;
+
+    Vector<CookieListItem> cookieList;
```
use Vector::map().

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    eventInit.deleted = WTFMove(cookieList);
+    eventInit.bubbles = false;
+    eventInit.cancelable = false;
+    queueTaskToDispatchEvent(*this, TaskSource::CookieStore, CookieChangeEvent::create(eventNames().changeEvent, WTFMove(eventInit)));
```
You need to set IsTrusted::Yes since this is a browser-initiated event.

cdumez replied:
You're using the wrong TaskSource, the spec says:
```
Queue a global task on the DOM manipulation task source given window to fire a change event named "change" with changes at window’s CookieStore.
```

cdumez replied:
Wrong task source.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    eventInit.changed = WTFMove(cookieList);
+    eventInit.bubbles = false;
+    eventInit.cancelable = false;
+    queueTaskToDispatchEvent(*this, TaskSource::CookieStore, CookieChangeEvent::create(eventNames().changeEvent, WTFMove(eventInit)));
```
You need to set IsTrusted::Yes since this is a browser-initiated event.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -329,4 +381,30 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
+    if (m_hasChangeEventListener)
+        cookieJar.addChangeListener(host, *this);
+    else
+        cookieJar.removeChangeListener(host, *this);
```
We should override ActiveDOMObject::stop() and call this there too if `m_hasChangeEventListener`

RupinMittal replied:
So we call it when we are removing the change listener? And this will remove the JS wrapper?

cdumez replied:
ActiveDOMObject::stop() gets called when the document is going away. No point is listening for cookie events past that point since you can't run script anymore.

cdumez commented on Source/WebCore/dom/TaskSource.h:
```diff
@@ -28,6 +28,7 @@
 namespace WebCore {
 
 enum class TaskSource : uint8_t {
+    CookieStore,
```
Not needed, couldn't see such thing in the spec.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+
+    CookieChangeEventInit eventInit;
+    eventInit.changed = WTFMove(cookieList);
+    eventInit.bubbles = false;
```
Not needed, this is the default

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    CookieChangeEventInit eventInit;
+    eventInit.changed = WTFMove(cookieList);
+    eventInit.bubbles = false;
+    eventInit.cancelable = false;
```
Not needed, this is the default

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+
+    CookieChangeEventInit eventInit;
+    eventInit.deleted = WTFMove(cookieList);
+    eventInit.bubbles = false;
```
Not needed, this is the default

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,6 +318,54 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    CookieChangeEventInit eventInit;
+    eventInit.deleted = WTFMove(cookieList);
+    eventInit.bubbles = false;
+    eventInit.cancelable = false;
```
Not needed, this is the default

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_basic.https.window-expected.txt:
```diff
@@ -1,5 +1,3 @@
-Harness Error (TIMEOUT), message = null
-
-TIMEOUT cookieStore fires change event for cookie set by cookieStore.set() Test timed out
+FAIL cookieStore fires change event for cookie set by cookieStore.set() assert_equals: expected [object "[object Object]"] but got [object "[object Object]"]
```
Would be nice to pass this "basic" test though.. :)

RupinMittal replied:
Yeah, that's true. I'll look into this and try and fix this in this patch itself

RupinMittal replied:
I'm a bit confused about this test. The line that is failing is `assert_equals(event.changed, event.changed);` And it's saying that `assert_equals: expected [object "[object Object]"] but got [object "[object Object]"]`. But aren't these the same variables that it's comparing?

cdumez replied:
If you comment out those 2 lines:
```
  assert_equals(event.changed, event.changed);
  assert_equals(event.deleted, event.deleted);
```

Is the test passing? If so, we can address I think you just need to add [CachedAttribute] in the IDL for `changed` & `deleted`.

`changed` & `deleted` are likely marked as [SameObject] in the idl but our implementation is currently slightly wrong and doesn't necessarily return the same objects.

RupinMittal replied:
Yes, if we comment out those lines, then the test passes. And yes, `changed` and `deleted` are marked as [SameObject]. What exactly does that mean? And what is the part of the implementation that is lightly wrong. Maybe I can try fixing it?

cdumez replied:
I told you how to fix it in my comment already.

RupinMittal replied:
Ah ok. I thought we needed [SameAttribute] eventually and there was some other implementation mistake and [CachedAttribute] would be a temporary fix for now. But I can change it to [CachedAttribute]. What does Cached Attribute mean here?

cdumez replied:
[CachedAttribute] is one of the ways to implement the SameObject behavior, it is the simplest one. A new data member is added to the JS wrapper to cache the value returned by the attribute the first time it is accessed. Then on follow-up calls, it will return this cached value.

You want to use `[CachedAttribute, SameObject]` in the IDL in your case. I believe it will fix your test and is a proper fix.

RupinMittal replied:
Ok that makes sense. It should be fixed now!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -36,13 +36,35 @@ namespace WebCore {
 
+    CookieListItem(const Cookie& cookie)
+    {
+        name = cookie.name;
```
Please use the initializer list.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -36,13 +36,35 @@ namespace WebCore {
+
     CookieListItem(Cookie&& cookie)
     {
         name = WTFMove(cookie.name);
```
Please use the initializer list.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+        return;
+
+    auto& document = *downcast<Document>(context);
+    UNUSED_PARAM(document);
```
Please omit.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+
+    auto& document = *downcast<Document>(context);
+    UNUSED_PARAM(document);
+    ASSERT(host == document.url().host().toString());
```
Use this instead:
```
ASSERT_UNUSED(document, host == document.url().host().toString());
```

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    Vector<CookieListItem> changedCookieList;
+    Vector<CookieListItem> deletedCookieList;
+    auto currentTime = WallTime::now().secondsSinceEpoch().milliseconds();
+    for (auto cookie : cookies) {
```
auto&

RupinMittal replied:
The compiler doesn't let me do this. It complains because with `auto`, the it treats the cookie as const and won't allow the below line `cookie.value = nullString();` to work. I have left it as just `auto` for now. Is this ok?

cdumez replied:
Ok

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+            changedCookieList.append(CookieListItem { cookie });
+    }
+
+    CookieChangeEventInit eventInit;
```
Can we create this before the for loop instead?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    for (auto cookie : cookies) {
+        if (cookie.expires && *cookie.expires < currentTime) {
+            cookie.value = nullString();
+            deletedCookieList.append(CookieListItem { cookie });
```
It you create eventInit before the `for` loop then you can append directly to `eventInit.deleted` instead of having a temporary vector.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+            cookie.value = nullString();
+            deletedCookieList.append(CookieListItem { cookie });
+        } else
+            changedCookieList.append(CookieListItem { cookie });
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, CookieChangeEvent::create(eventNames().changeEvent, WTFMove(eventInit), CookieChangeEvent::IsTrusted::Yes));
+}
+
+void CookieStore::cookiesDeleted(const String& host, const Vector<Cookie>& cookies)
```
Same comments as in the function above.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+        return;
+
+    auto* context = scriptExecutionContext();
+    if (!context)
```
Hmm, Actually I believe this is always be null when `stop()` gets called I believe. I think this shows that we should store both the `CookieJar` and the `host` as data members.

RupinMittal commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt:
```diff
@@ -1,8 +1,6 @@
+PASS document.cookie set/overwrite/delete observed by CookieStore
+PASS CookieStore set/overwrite/delete observed by document.cookie
+FAIL CookieStore agrees with document.cookie on encoding non-ASCII cookies assert_equals: Cookie we wrote using document.cookie in cookie jar expected (string) "DOCUMENT-🍪=🔵" but got (undefined) undefined
+PASS document.cookie agrees with CookieStore on encoding non-ASCII cookies
```
This test seems to be flaky. But I'm not quite sure how to fix it.

RupinMittal replied:
I have marked the test as flaky in TestExpectations and filed a bug to be revisited later: https://bugs.webkit.org/show_bug.cgi?id=260103

cdumez replied:
File a bug, mark as flaky in test expectations and investigate later.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -36,15 +36,35 @@ namespace WebCore {
-        path = WTFMove(cookie.path);
-        expires = cookie.expires.has_value() ? std::optional { *cookie.expires } : std::nullopt;
-        secure = cookie.secure;
+        switch (cookie.sameSite) {
```
Should be moved to an inline function in the header to avoid code duplication.

RupinMittal replied:
Since this new constructor is not needed, I have removed it and not inlined this block to a new function. Is this ok?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -36,15 +36,35 @@ namespace WebCore {
+        , expires(cookie.expires)
+        , secure(cookie.secure)
+    {
         switch (cookie.sameSite) {
```
See duplication here.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -62,6 +66,15 @@ Ref<CookieStore> CookieStore::create(Document* document)
+    if (!page)
+        return;
+
+    m_cookieJar = WeakPtr { page->cookieJar() };
```
Should be no need for an explicit `WeakPtr { }`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +319,69 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+            cookie.value = nullString();
+            eventInit.deleted.append(CookieListItem { cookie });
+        } else
+            eventInit.changed.append(CookieListItem { cookie });
```
`WTFMove(cookie)`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +319,69 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    for (auto cookie : cookies) {
+        if (cookie.expires && *cookie.expires < currentTime) {
+            cookie.value = nullString();
+            eventInit.deleted.append(CookieListItem { cookie });
```
`WTFMove(cookie)`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +319,69 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+    CookieChangeEventInit eventInit;
+    eventInit.deleted = cookies.map([](auto cookie) {
+        cookie.value = nullString();
+        return CookieListItem { cookie };
```
`WTFMove(cookie)`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -36,15 +36,35 @@ namespace WebCore {
     CookieListItem() = default;
 
-    CookieListItem(Cookie&& cookie)
+    CookieListItem(const Cookie& cookie)
```
Please double check that this new constructor is still needed after the changes I am suggesting below.

RupinMittal replied:
It's not needed anymore so I have removed and not inlined the body of the constructor to a new function

cdumez commented on LayoutTests/TestExpectations:
```diff
@@ -1826,6 +1826,7 @@ imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/t
 imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/audio_loop_base.html [ Failure Pass ]
 imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/networkState_during_progress.html [ Failure Pass ]
 imported/w3c/web-platform-tests/html/semantics/embedded-content/the-object-element/object-fallback-failed-cross-origin-navigation.sub.html [ Failure Pass ]
+imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window.html [ Pass Failure ]
```
Where is the bug link? Something like `webkit.org/b/186045`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +319,69 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+void CookieStore::cookiesDeleted(const String& host, const Vector<Cookie>& cookies)
+{
+    ASSERT_UNUSED(host, host == m_host);
+    if (!m_hasChangeEventListener)
```
Should this be an assertion? Why would we be getting notifications if we don't have a change event listener?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -314,11 +319,69 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
+void CookieStore::cookiesAdded(const String& host, const Vector<Cookie>& cookies)
+{
+    ASSERT_UNUSED(host, host == m_host);
+    if (!m_hasChangeEventListener)
```
Shouldn't this be an assertion? I don't think we should be getting cookie notifications if we don't have a listener.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -329,4 +392,32 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
+    if (!page)
+        return;
+
+    auto& cookieJar = page->cookieJar();
```
This function needs updating now that you store the cookie jar and host as data members.

cdumez reviewed (approved):
LGTM assuming the bots are happy.

RupinMittal committed:
Add the change event for the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=260028

Reviewed by Chris Dumez.

CookieStore now subclasses CookieChangeListener so that when an
event listener changes, it can register (or unregister) itself
as a listener. If it is registered and a cookie changed, the
relevant function (cookiesAdded or cookiesDeleted) will be called
to fire a change event.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_basic.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_delete.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_overwrite.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/httponly_cookies.https.window-expected.txt:
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt: Added.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt: Added.
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.idl:
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
(WebCore::CookieListItem::CookieListItem):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::cookiesAdded):
(WebCore::CookieStore::cookiesDeleted):
(WebCore::CookieStore::stop):
(WebCore::CookieStore::virtualHasPendingActivity const):
(WebCore::CookieStore::eventListenersDidChange):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/loader/CookieJar.h:

Canonical link: https://commits.webkit.org/266850@main
